llvm/test/CodeGen/StubV/slt.mir

```diff
@@ -0,0 +1,77 @@
+# RUN: llc %s -mtriple=stubv32 -start-after=unpack-mi-bundles -o - | FileCheck %s
+
+--- |
+  ; ModuleID = '<stdin>'
+  source_filename = "<stdin>"
+  target datalayout = "e-m:e-p:32:32-i64:64-n32-S128"
+  target triple = "stubv32"
+
+  define i32 @foo(i32 %a, i32 %b) {
+  ; CHECK-LABEL: foo:
+  ; CHECK:       # %bb.0:
+  ; CHECK-NEXT:    slt a0, a0, a1
+  ; CHECK-NEXT:    ret
+    %c = slt i32 %a, %b
```

Comment:
```suggestion
    %c = icmp slt i32 %a, %b
```

The LLVM IR would look more like this I think?

Reply:
Which actually opens up a general question of how this IR is used.  I thought it was still processed, if so then this test would not pass as is (not because the CHECKs fail but because the IR doesn't parse).

Or is the IR only used by the generation script?

Reply:
Yep, you are right. I forgot to check how it would be represented in LLVM IR :P 

There is definitely some very strange behaviour going on which I dismissed initially. I swear at the end it was passing via `ninja check-llvm-codegen-stubv` and failing when running `llc %s -mtriple=stubv32 -start-after=unpack-mi-bundles -o -` manually. 

Currently fails using both commands as it currently is. Which is good. Though, perhaps more concerning is this passes the filecheck when `slt` is replaced with `add` in line 14 which seems to suggest filecheck is not doing anything for some reason.

Reply:
Lol, what is going on. I have this IR:

```
--- |
  ; ModuleID = '<stdin>'
  source_filename = "<stdin>"
  target datalayout = "e-m:e-p:32:32-i64:64-n32-S128"
  target triple = "stubv32"

  define i32 @foo(i32 %a, i32 %b) {
  ; CHECK-LABEL: foo:
  ; CHECK:       # %bb.0:
  ; CHECK-NEXT:    slt a0, a0, a1
  ; CHECK-NEXT:    ret
    %c = add i32 %a, %b
    ret i32 %c
  }

...
```

Yet when running `/home/duncan/documents/git/stubv/llvm-project/build/bin/llc /home/duncan/documents/git/stubv/llvm-project/llvm/test/CodeGen/StubV/slt.mir -mtriple=stubv32 -start-after=unpack-mi-bundles -o -`

I get this:

```
        .text
        .file   "<stdin>"
        .globl  foo                             # -- Begin function foo
        .p2align        2
        .type   foo,@function
foo:                                    # @foo
        .cfi_startproc
# %bb.0:
        slt     a0, a0, a1
        ret
.Lfunc_end0:
        .size   foo, .Lfunc_end0-foo
        .cfi_endproc
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```
No `add` in sight. Only slt.

Reply:
You get the `slt` StubV instruction even when you have `add` in the LLVM IR because the test is actually processing the MIR.  Which still has the `renamable $x10 = SLT killed renamable $x10, killed renamable $x11` line in it, i.e. that MIR `SLT` is being translated for the StubV `slt` instruction, not the LLVM IR.

But changing the LLVM IR to something illegal still fails the test because it still seems to be parsed, just not used.

I'm stating this because I went through a similar sequence of thoughts when deliberately breaking my own test (just as you have been doing).  Perhaps others could explain *why* it works this way.

Reply:
LLVM's Machine IR (MIR) is not a completely independent representation. It still relies on values in the LLVM IR for certain things. That's why a `MachineFunction`points to a `Function` (notice that `name: foo` is required, and a `@foo` must exist in the IR), and a `MachineBasicBlock` points to a ` BasicBlock` (`bb.0 (%ir-block.0)`), etc.

The serialized MIR has an embedded LLVM IR module so it can reference IR values when needed.

See a more explicit example of the fact that alias analysis always takes place on LLVM IR values - even in MIR - from this AMDGPU snippet:

```
  %scevgep = getelementptr i8, ptr getelementptr inbounds ([3 x i8], ptr @c, i64 0, i64 1), i64 0

...
    %4:gpr32 = LDRBBui killed %3, 1 :: (load (s8) from %ir.scevgep)
```

As @alycm pointed out, it happens this test doesn't really *need* the body of the function in the LLVM IR because it's starting after instruction selection has already taken place (`-start-after=unpack-mi-bundles`)

It could just as well be:

``` llvm
  define void @foo() {
    ret void
  }
```

but I'd argue that's _more_ confusing, at least once you understand the relationship between MIR and IR.

Hope that helps!